From 517295c9eea8a90ffd276399e503a9d57b154fce Mon Sep 17 00:00:00 2001
From: Johan Cwiklinski <johan@x-tnd.be>
Date: Thu, 8 Feb 2024 11:11:35 +0100
Subject: [PATCH] Fix network equipment memory conversion

---
 lib/php/Converter.php | 13 ++++++-------
 1 file changed, 6 insertions(+), 7 deletions(-)

diff --git a/lib/php/Converter.php b/lib/php/Converter.php
index 6e91274..519d221 100644
--- a/lib/php/Converter.php
+++ b/lib/php/Converter.php
@@ -936,13 +936,12 @@ class Converter
             }
         }
         if (isset($data['content']['network_device'])) {
-            foreach ($data['content']['network_device'] as &$netdev) {
-                if (isset($netdev['memory'])) {
-                    $netdev['memory'] = $this->convertMemory($netdev['memory']);
-                }
-                if (isset($netdev['ram'])) {
-                    $netdev['ram'] = $this->convertMemory($netdev['ram']);
-                }
+            $netdev = &$data['content']['network_device'];
+            if (isset($netdev['memory'])) {
+                $netdev['memory'] = $this->convertMemory($netdev['memory']);
+            }
+            if (isset($netdev['ram'])) {
+                $netdev['ram'] = $this->convertMemory($netdev['ram']);
             }
         }
 
-- 
2.43.0

