diff --git a/inventory.schema.json b/inventory.schema.json
index 609dbb0..fcf8d86 100644
--- a/inventory.schema.json
+++ b/inventory.schema.json
@@ -2904,7 +2904,7 @@
                   "type": "string",
                   "title": "PDU type"
                 },
-                "plugs": {
+                "plug": {
                   "title": "List of plugs of pdu",
                   "type": "array",
                   "items": {
diff --git a/lib/php/Converter.php b/lib/php/Converter.php
index 4f9d759..a2a7ddc 100644
--- a/lib/php/Converter.php
+++ b/lib/php/Converter.php
@@ -1440,6 +1440,9 @@ public function convertNetworkInventory(array $data): array
                     break;
                 case "pdu":
                     // put pdu into network_device
+                    if (isset($device_data['plug']) && !array_is_list($device_data['plug'])) {
+                        $device_data['plug'] = [$device_data['plug']];
+                    }
                     $data['content']['network_device'][$key] = $device_data;
                     break;
                 default:
