diff --git a/src/Formatter.php b/src/Formatter.php
index 9e06569..ddd64e7 100644
--- a/src/Formatter.php
+++ b/src/Formatter.php
@@ -331,9 +331,7 @@ abstract class Formatter
             foreach ($lines as $line) {
                 // If the written line starts is empty, adding a mapping would add it for
                 // a non-existent column as we are at the end of the line
-                if ($line !== '') {
-                    assert($this->currentBlock->sourceLine !== null);
-                    assert($this->currentBlock->sourceName !== null);
+                if ($line !== '' && $this->currentBlock->sourceLine !== null && $this->currentBlock->sourceName !== null) {
                     $this->sourceMapGenerator->addMapping(
                         $this->currentLine,
                         $this->currentColumn,
@@ -348,9 +346,7 @@ abstract class Formatter
                 $this->currentColumn = 0;
             }
 
-            if ($lastLine !== '') {
-                assert($this->currentBlock->sourceLine !== null);
-                assert($this->currentBlock->sourceName !== null);
+            if ($lastLine !== '' && $this->currentBlock->sourceLine !== null && $this->currentBlock->sourceName !== null) {
                 $this->sourceMapGenerator->addMapping(
                     $this->currentLine,
                     $this->currentColumn,
