diff --git a/StringExtension.php b/StringExtension.php
index 7b5d0049..2e827af0 100644
--- a/StringExtension.php
+++ b/StringExtension.php
@@ -22,7 +22,7 @@ final class StringExtension extends AbstractExtension
 {
     private $slugger;
 
-    public function __construct(SluggerInterface $slugger = null)
+    public function __construct(?SluggerInterface $slugger = null)
     {
         $this->slugger = $slugger ?: new AsciiSlugger();
     }
