diff -u a/src/Cache/ChainCache.php b/src/Cache/ChainCache.php
--- a/src/Cache/ChainCache.php
+++ b/src/Cache/ChainCache.php
@@ -21,12 +21,15 @@
  */
 final class ChainCache implements CacheInterface, RemovableCacheInterface
 {
+    private iterable $caches;
+
     /**
      * @param iterable<CacheInterface> $caches The ordered list of caches used to store and fetch cached items
      */
     public function __construct(
-        private iterable $caches,
+        iterable $caches
     ) {
+        $this->caches = $caches;
     }
 
     public function generateKey(string $name, string $className): string
diff -u a/src/Compiler.php b/src/Compiler.php
--- a/src/Compiler.php
+++ b/src/Compiler.php
@@ -29,9 +29,12 @@
     private $didUseEcho = false;
     private $didUseEchoStack = [];
 
+    private Environment $env;
+
     public function __construct(
-        private Environment $env,
+        Environment $env
     ) {
+        $this->env = $env;
     }
 
     public function getEnvironment(): Environment
@@ -74,7 +77,7 @@
             $node->compile($this);
 
             if ($this->didUseEcho) {
-                trigger_deprecation('twig/twig', '3.9', 'Using "%s" is deprecated, use "yield" instead in "%s", then flag the class with #[\Twig\Attribute\YieldReady].', $this->didUseEcho, $node::class);
+                trigger_deprecation('twig/twig', '3.9', 'Using "%s" is deprecated, use "yield" instead in "%s", then flag the class with #[\Twig\Attribute\YieldReady].', $this->didUseEcho, get_class($node));
             }
 
             return $this;
@@ -99,7 +102,7 @@
             $node->compile($this);
 
             if ($this->didUseEcho) {
-                trigger_deprecation('twig/twig', '3.9', 'Using "%s" is deprecated, use "yield" instead in "%s", then flag the class with #[\Twig\Attribute\YieldReady].', $this->didUseEcho, $node::class);
+                trigger_deprecation('twig/twig', '3.9', 'Using "%s" is deprecated, use "yield" instead in "%s", then flag the class with #[\Twig\Attribute\YieldReady].', $this->didUseEcho, get_class($node));
             }
 
             return $this;
diff -u a/src/DeprecatedCallableInfo.php b/src/DeprecatedCallableInfo.php
--- a/src/DeprecatedCallableInfo.php
+++ b/src/DeprecatedCallableInfo.php
@@ -19,13 +19,24 @@
     private string $type;
     private string $name;
 
+    private string $package;
+    private string $version;
+    private ?string $altName;
+    private ?string $altPackage;
+    private ?string $altVersion;
+
     public function __construct(
-        private string $package,
-        private string $version,
-        private ?string $altName = null,
-        private ?string $altPackage = null,
-        private ?string $altVersion = null,
+        string $package,
+        string $version,
+        ?string $altName = null,
+        ?string $altPackage = null,
+        ?string $altVersion = null
     ) {
+        $this->package = $package;
+        $this->version = $version;
+        $this->altName = $altName;
+        $this->altPackage = $altPackage;
+        $this->altVersion = $altVersion;
     }
 
     public function setType(string $type): void
diff -u a/src/ExpressionParser/AbstractExpressionParser.php b/src/ExpressionParser/AbstractExpressionParser.php
--- a/src/ExpressionParser/AbstractExpressionParser.php
+++ b/src/ExpressionParser/AbstractExpressionParser.php
@@ -15,7 +15,14 @@
 {
     public function __toString(): string
     {
-        return \sprintf('%s(%s)', ExpressionParserType::getType($this)->value, $this->getName());
+        if ($this instanceof PrefixExpressionParserInterface) {
+            $type = 'prefix';
+        }
+        if ($this instanceof InfixExpressionParserInterface) {
+            $type = 'infix';
+        }
+
+        return \sprintf('%s(%s)', $type, $this->getName());
     }
 
     public function getPrecedenceChange(): ?PrecedenceChange
diff -u a/src/ExpressionParser/ExpressionParsers.php b/src/ExpressionParser/ExpressionParsers.php
--- a/src/ExpressionParser/ExpressionParsers.php
+++ b/src/ExpressionParser/ExpressionParsers.php
@@ -46,7 +46,7 @@
      *
      * @return $this
      */
-    public function add(array $parsers): static
+    public function add(array $parsers)/*: static*/
     {
         foreach ($parsers as $parser) {
             if ($parser->getPrecedence() > 512 || $parser->getPrecedence() < 0) {
@@ -54,7 +54,7 @@
                 // throw new \InvalidArgumentException(\sprintf('Precedence for "%s" must be between 0 and 512, got %d.', $parser->getName(), $parser->getPrecedence()));
             }
             $interface = $parser instanceof PrefixExpressionParserInterface ? PrefixExpressionParserInterface::class : InfixExpressionParserInterface::class;
-            $this->parsersByClass[$parser::class] = $parser;
+            $this->parsersByClass[get_class($parser)] = $parser;
             foreach (self::getOperatorTokensFor($parser) as $token) {
                 $this->parsersByName[$interface][$token] = $parser;
             }
@@ -128,7 +128,7 @@
                         if (!isset($this->precedenceChanges[$e])) {
                             $this->precedenceChanges[$e] = [];
                         }
-                        $this->precedenceChanges[$e][] = $ep;
+                        @$this->precedenceChanges[$e][] = $ep;
                     }
                 }
             }
@@ -148,7 +148,7 @@
             return $parser->getOperatorTokens();
         }
 
-        trigger_deprecation('twig/twig', '3.24', 'Not implementing the "getOperatorTokens()" method in "%s" is deprecated. This method will be part of the "%s" interface in 4.0.', $parser::class, ExpressionParserInterface::class);
+        trigger_deprecation('twig/twig', '3.24', 'Not implementing the "getOperatorTokens()" method in "%s" is deprecated. This method will be part of the "%s" interface in 4.0.', get_class($parser), ExpressionParserInterface::class);
 
         return [$parser->getName(), ...$parser->getAliases()];
     }
diff -u a/src/ExpressionParser/Infix/ArgumentsTrait.php b/src/ExpressionParser/Infix/ArgumentsTrait.php
--- a/src/ExpressionParser/Infix/ArgumentsTrait.php
+++ b/src/ExpressionParser/Infix/ArgumentsTrait.php
@@ -64,7 +64,7 @@
                 $value = $value->getNode('right');
             } elseif (($token = $stream->nextIf(Token::OPERATOR_TYPE, '=')) || ($token = $stream->nextIf(Token::PUNCTUATION_TYPE, ':'))) {
                 if (!$value instanceof ContextVariable) {
-                    throw new SyntaxError(\sprintf('A parameter name must be a string, "%s" given.', $value::class), $token->getLine(), $stream->getSourceContext());
+                    throw new SyntaxError(\sprintf('A parameter name must be a string, "%s" given.', get_class($value)), $token->getLine(), $stream->getSourceContext());
                 }
                 $name = $value->getAttribute('name');
                 $value = $parser->parseExpression();
diff -u a/src/ExpressionParser/Infix/ArrowExpressionParser.php b/src/ExpressionParser/Infix/ArrowExpressionParser.php
--- a/src/ExpressionParser/Infix/ArrowExpressionParser.php
+++ b/src/ExpressionParser/Infix/ArrowExpressionParser.php
@@ -13,7 +13,6 @@
 
 use Twig\ExpressionParser\AbstractExpressionParser;
 use Twig\ExpressionParser\ExpressionParserDescriptionInterface;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\ExpressionParser\InfixExpressionParserInterface;
 use Twig\Node\Expression\AbstractExpression;
 use Twig\Node\Expression\ArrowFunctionExpression;
@@ -46,8 +45,8 @@
         return 250;
     }
 
-    public function getAssociativity(): InfixAssociativity
+    public function getAssociativity(): int
     {
-        return InfixAssociativity::Left;
+        return 1;
     }
 }
diff -u a/src/ExpressionParser/Infix/AssignmentExpressionParser.php b/src/ExpressionParser/Infix/AssignmentExpressionParser.php
--- a/src/ExpressionParser/Infix/AssignmentExpressionParser.php
+++ b/src/ExpressionParser/Infix/AssignmentExpressionParser.php
@@ -12,7 +12,6 @@
 namespace Twig\ExpressionParser\Infix;
 
 use Twig\Error\SyntaxError;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\Node\Expression\AbstractExpression;
 use Twig\Node\Expression\ArrayExpression;
 use Twig\Node\Expression\Binary\AbstractBinary;
@@ -29,9 +28,9 @@
 class AssignmentExpressionParser extends BinaryOperatorExpressionParser
 {
     public function __construct(
-        string $name,
+        string $name
     ) {
-        parent::__construct(SetBinary::class, $name, 0, InfixAssociativity::Right);
+        parent::__construct(SetBinary::class, $name, 0, 2);
     }
 
     /**
@@ -40,12 +39,11 @@
     public function parse(Parser $parser, AbstractExpression $left, Token $token): AbstractExpression
     {
         if (!$left instanceof ContextVariable && !$left instanceof ArrayExpression) {
-            throw new SyntaxError(\sprintf('Cannot assign to "%s", only variables can be assigned.', $left::class), $token->getLine(), $parser->getStream()->getSourceContext());
+            throw new SyntaxError(\sprintf('Cannot assign to "%s", only variables can be assigned.', get_class($left)), $token->getLine(), $parser->getStream()->getSourceContext());
         }
-        $right = $parser->parseExpression(InfixAssociativity::Left === $this->getAssociativity() ? $this->getPrecedence() + 1 : $this->getPrecedence());
-        $right = match ($this->getName()) {
-            '=' => $right,
-            default => throw new \LogicException(\sprintf('Unknown operator: %s.', $this->getName())),
+        $right = $parser->parseExpression(1 === $this->getAssociativity() ? $this->getPrecedence() + 1 : $this->getPrecedence());
+        if ($this->getName() !== '=') {
+            throw new \LogicException(\sprintf('Unknown operator: %s.', $this->getName()));
         };
 
         if ($left instanceof ArrayExpression) {
diff -u a/src/ExpressionParser/Infix/BinaryOperatorExpressionParser.php b/src/ExpressionParser/Infix/BinaryOperatorExpressionParser.php
--- a/src/ExpressionParser/Infix/BinaryOperatorExpressionParser.php
+++ b/src/ExpressionParser/Infix/BinaryOperatorExpressionParser.php
@@ -13,7 +13,6 @@
 
 use Twig\ExpressionParser\AbstractExpressionParser;
 use Twig\ExpressionParser\ExpressionParserDescriptionInterface;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\ExpressionParser\InfixExpressionParserInterface;
 use Twig\ExpressionParser\PrecedenceChange;
 use Twig\Node\Expression\AbstractExpression;
@@ -26,16 +25,31 @@
  */
 class BinaryOperatorExpressionParser extends AbstractExpressionParser implements InfixExpressionParserInterface, ExpressionParserDescriptionInterface
 {
+    private string $nodeClass;
+    private string $name;
+    private int $precedence;
+    private int $associativity;
+    private ?PrecedenceChange $precedenceChange;
+    private ?string $description;
+    private array $aliases;
+
     public function __construct(
         /** @var class-string<AbstractBinary> */
-        private string $nodeClass,
-        private string $name,
-        private int $precedence,
-        private InfixAssociativity $associativity = InfixAssociativity::Left,
-        private ?PrecedenceChange $precedenceChange = null,
-        private ?string $description = null,
-        private array $aliases = [],
+        string $nodeClass,
+        string $name,
+        int $precedence,
+        int $associativity = 1,
+        ?PrecedenceChange $precedenceChange = null,
+        ?string $description = null,
+        array $aliases = []
     ) {
+        $this->nodeClass = $nodeClass;
+        $this->name = $name;
+        $this->precedence = $precedence;
+        $this->associativity = $associativity;
+        $this->precedenceChange = $precedenceChange;
+        $this->description = $description;
+        $this->aliases = $aliases;
     }
 
     /**
@@ -43,12 +57,13 @@
      */
     public function parse(Parser $parser, AbstractExpression $left, Token $token): AbstractExpression
     {
-        $right = $parser->parseExpression(InfixAssociativity::Left === $this->getAssociativity() ? $this->getPrecedence() + 1 : $this->getPrecedence());
+        $right = $parser->parseExpression(1 === $this->getAssociativity() ? $this->getPrecedence() + 1 : $this->getPrecedence());
 
-        return new ($this->nodeClass)($left, $right, $token->getLine());
+        $nodeClass = $this->nodeClass;
+        return new $nodeClass($left, $right, $token->getLine());
     }
 
-    public function getAssociativity(): InfixAssociativity
+    public function getAssociativity(): int
     {
         return $this->associativity;
     }
diff -u a/src/ExpressionParser/Infix/ConditionalTernaryExpressionParser.php b/src/ExpressionParser/Infix/ConditionalTernaryExpressionParser.php
--- a/src/ExpressionParser/Infix/ConditionalTernaryExpressionParser.php
+++ b/src/ExpressionParser/Infix/ConditionalTernaryExpressionParser.php
@@ -13,7 +13,6 @@
 
 use Twig\ExpressionParser\AbstractExpressionParser;
 use Twig\ExpressionParser\ExpressionParserDescriptionInterface;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\ExpressionParser\InfixExpressionParserInterface;
 use Twig\Node\Expression\AbstractExpression;
 use Twig\Node\Expression\ConstantExpression;
@@ -55,8 +54,8 @@
         return 0;
     }
 
-    public function getAssociativity(): InfixAssociativity
+    public function getAssociativity(): int
     {
-        return InfixAssociativity::Left;
+        return 1;
     }
 }
diff -u a/src/ExpressionParser/Infix/DotExpressionParser.php b/src/ExpressionParser/Infix/DotExpressionParser.php
--- a/src/ExpressionParser/Infix/DotExpressionParser.php
+++ b/src/ExpressionParser/Infix/DotExpressionParser.php
@@ -14,7 +14,6 @@
 use Twig\Error\SyntaxError;
 use Twig\ExpressionParser\AbstractExpressionParser;
 use Twig\ExpressionParser\ExpressionParserDescriptionInterface;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\ExpressionParser\InfixExpressionParserInterface;
 use Twig\Lexer;
 use Twig\Node\Expression\AbstractExpression;
@@ -101,8 +100,8 @@
         return 512;
     }
 
-    public function getAssociativity(): InfixAssociativity
+    public function getAssociativity(): int
     {
-        return InfixAssociativity::Left;
+        return 1;
     }
 }
diff -u a/src/ExpressionParser/Infix/FilterExpressionParser.php b/src/ExpressionParser/Infix/FilterExpressionParser.php
--- a/src/ExpressionParser/Infix/FilterExpressionParser.php
+++ b/src/ExpressionParser/Infix/FilterExpressionParser.php
@@ -14,7 +14,6 @@
 use Twig\Attribute\FirstClassTwigCallableReady;
 use Twig\ExpressionParser\AbstractExpressionParser;
 use Twig\ExpressionParser\ExpressionParserDescriptionInterface;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\ExpressionParser\InfixExpressionParserInterface;
 use Twig\ExpressionParser\PrecedenceChange;
 use Twig\Node\EmptyNode;
@@ -46,12 +45,14 @@
 
         $filter = $parser->getFilter($token->getValue(), $line);
 
+        $class = $filter->getNodeClass();
+
         $ready = true;
-        if (!isset($this->readyNodes[$class = $filter->getNodeClass()])) {
+        if (\PHP_VERSION_ID >= 80000 && !isset($this->readyNodes[$class])) {
             $this->readyNodes[$class] = (bool) (new \ReflectionClass($class))->getConstructor()->getAttributes(FirstClassTwigCallableReady::class);
         }
 
-        if (!$ready = $this->readyNodes[$class]) {
+        if (\PHP_VERSION_ID >= 80000 && !$ready = $this->readyNodes[$class]) {
             trigger_deprecation('twig/twig', '3.12', 'Twig node "%s" is not marked as ready for passing a "TwigFilter" in the constructor instead of its name; please update your code and then add #[FirstClassTwigCallableReady] attribute to the constructor.', $class);
         }
 
@@ -78,8 +79,8 @@
         return new PrecedenceChange('twig/twig', '3.21', 300);
     }
 
-    public function getAssociativity(): InfixAssociativity
+    public function getAssociativity(): int
     {
-        return InfixAssociativity::Left;
+        return 1;
     }
 }
diff -u a/src/ExpressionParser/Infix/FunctionExpressionParser.php b/src/ExpressionParser/Infix/FunctionExpressionParser.php
--- a/src/ExpressionParser/Infix/FunctionExpressionParser.php
+++ b/src/ExpressionParser/Infix/FunctionExpressionParser.php
@@ -15,7 +15,6 @@
 use Twig\Error\SyntaxError;
 use Twig\ExpressionParser\AbstractExpressionParser;
 use Twig\ExpressionParser\ExpressionParserDescriptionInterface;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\ExpressionParser\InfixExpressionParserInterface;
 use Twig\Node\EmptyNode;
 use Twig\Node\Expression\AbstractExpression;
@@ -63,11 +62,14 @@
             return $node;
         }
 
-        if (!isset($this->readyNodes[$class = $function->getNodeClass()])) {
+        $class = $function->getNodeClass();
+
+        $ready = true;
+        if (\PHP_VERSION_ID >= 80000 && !isset($this->readyNodes[$class])) {
             $this->readyNodes[$class] = (bool) (new \ReflectionClass($class))->getConstructor()->getAttributes(FirstClassTwigCallableReady::class);
         }
 
-        if (!$ready = $this->readyNodes[$class]) {
+        if (\PHP_VERSION_ID >= 80000 && !$ready = $this->readyNodes[$class]) {
             trigger_deprecation('twig/twig', '3.12', 'Twig node "%s" is not marked as ready for passing a "TwigFunction" in the constructor instead of its name; please update your code and then add #[FirstClassTwigCallableReady] attribute to the constructor.', $class);
         }
 
@@ -89,8 +91,8 @@
         return 512;
     }
 
-    public function getAssociativity(): InfixAssociativity
+    public function getAssociativity(): int
     {
-        return InfixAssociativity::Left;
+        return 1;
     }
 }
diff -u a/src/ExpressionParser/Infix/IsExpressionParser.php b/src/ExpressionParser/Infix/IsExpressionParser.php
--- a/src/ExpressionParser/Infix/IsExpressionParser.php
+++ b/src/ExpressionParser/Infix/IsExpressionParser.php
@@ -14,7 +14,6 @@
 use Twig\Attribute\FirstClassTwigCallableReady;
 use Twig\ExpressionParser\AbstractExpressionParser;
 use Twig\ExpressionParser\ExpressionParserDescriptionInterface;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\ExpressionParser\InfixExpressionParserInterface;
 use Twig\Node\Expression\AbstractExpression;
 use Twig\Node\Expression\ArrayExpression;
@@ -50,12 +49,14 @@
             $expr = new MacroReferenceExpression($alias['node']->getNode('var'), $alias['name'], new ArrayExpression([], $expr->getTemplateLine()), $expr->getTemplateLine());
         }
 
+        $class = $test->getNodeClass();
+
         $ready = $test instanceof TwigTest;
-        if (!isset($this->readyNodes[$class = $test->getNodeClass()])) {
+        if (\PHP_VERSION_ID >= 80000 && !isset($this->readyNodes[$class])) {
             $this->readyNodes[$class] = (bool) (new \ReflectionClass($class))->getConstructor()->getAttributes(FirstClassTwigCallableReady::class);
         }
 
-        if (!$ready = $this->readyNodes[$class]) {
+        if (\PHP_VERSION_ID >= 80000 && !$ready = $this->readyNodes[$class]) {
             trigger_deprecation('twig/twig', '3.12', 'Twig node "%s" is not marked as ready for passing a "TwigTest" in the constructor instead of its name; please update your code and then add #[FirstClassTwigCallableReady] attribute to the constructor.', $class);
         }
 
@@ -77,8 +78,8 @@
         return 'Twig tests';
     }
 
-    public function getAssociativity(): InfixAssociativity
+    public function getAssociativity(): int
     {
-        return InfixAssociativity::Left;
+        return 1;
     }
 }
diff -u a/src/ExpressionParser/Infix/SquareBracketExpressionParser.php b/src/ExpressionParser/Infix/SquareBracketExpressionParser.php
--- a/src/ExpressionParser/Infix/SquareBracketExpressionParser.php
+++ b/src/ExpressionParser/Infix/SquareBracketExpressionParser.php
@@ -13,7 +13,6 @@
 
 use Twig\ExpressionParser\AbstractExpressionParser;
 use Twig\ExpressionParser\ExpressionParserDescriptionInterface;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\ExpressionParser\InfixExpressionParserInterface;
 use Twig\Node\Expression\AbstractExpression;
 use Twig\Node\Expression\ArrayExpression;
@@ -57,7 +56,9 @@
 
             $filter = $parser->getFilter('slice', $token->getLine());
             $arguments = new Nodes([$attribute, $length]);
-            $filter = new ($filter->getNodeClass())($expr, $filter, $arguments, $token->getLine());
+
+            $nodeClass = $filter->getNodeClass();
+            $filter = new $nodeClass($expr, $filter, $arguments, $token->getLine());
 
             $stream->expect(Token::PUNCTUATION_TYPE, ']');
 
@@ -84,8 +85,8 @@
         return 512;
     }
 
-    public function getAssociativity(): InfixAssociativity
+    public function getAssociativity(): int
     {
-        return InfixAssociativity::Left;
+        return 1;
     }
 }
diff -u a/src/ExpressionParser/InfixExpressionParserInterface.php b/src/ExpressionParser/InfixExpressionParserInterface.php
--- a/src/ExpressionParser/InfixExpressionParserInterface.php
+++ b/src/ExpressionParser/InfixExpressionParserInterface.php
@@ -23,5 +23,5 @@
      */
     public function parse(Parser $parser, AbstractExpression $left, Token $token): AbstractExpression;
 
-    public function getAssociativity(): InfixAssociativity;
+    public function getAssociativity(): int;
 }
diff -u a/src/ExpressionParser/PrecedenceChange.php b/src/ExpressionParser/PrecedenceChange.php
--- a/src/ExpressionParser/PrecedenceChange.php
+++ b/src/ExpressionParser/PrecedenceChange.php
@@ -18,11 +18,18 @@
  */
 class PrecedenceChange
 {
+    private string $package;
+    private string $version;
+    private int $newPrecedence;
+
     public function __construct(
-        private string $package,
-        private string $version,
-        private int $newPrecedence,
+        string $package,
+        string $version,
+        int $newPrecedence
     ) {
+        $this->package = $package;
+        $this->version = $version;
+        $this->newPrecedence = $newPrecedence;
     }
 
     public function getPackage(): string
diff -u a/src/ExpressionParser/Prefix/GroupingExpressionParser.php b/src/ExpressionParser/Prefix/GroupingExpressionParser.php
--- a/src/ExpressionParser/Prefix/GroupingExpressionParser.php
+++ b/src/ExpressionParser/Prefix/GroupingExpressionParser.php
@@ -59,7 +59,7 @@
             throw new SyntaxError('A list of variables must be followed by an arrow.', $stream->getCurrent()->getLine(), $stream->getSourceContext());
         }
 
-        return new ListExpression(array_map(self::toAssignContextVariable(...), $names), $token->getLine());
+        return new ListExpression(array_map([self::class, 'toAssignContextVariable'], $names), $token->getLine());
     }
 
     private static function toAssignContextVariable(AbstractExpression $expr): AssignContextVariable
diff -u a/src/ExpressionParser/Prefix/LiteralExpressionParser.php b/src/ExpressionParser/Prefix/LiteralExpressionParser.php
--- a/src/ExpressionParser/Prefix/LiteralExpressionParser.php
+++ b/src/ExpressionParser/Prefix/LiteralExpressionParser.php
@@ -67,11 +67,8 @@
 
             case $token->test(Token::PUNCTUATION_TYPE):
                 // In 4.0, we should always return the node or throw an error for default
-                if ($node = match ($token->getValue()) {
-                    '{' => $this->parseMappingExpression($parser),
-                    default => null,
-                }) {
-                    return $node;
+                if ($token->getValue() === '{') {
+                    return $this->parseMappingExpression($parser);
                 }
 
                 // no break
diff -u a/src/ExpressionParser/Prefix/UnaryOperatorExpressionParser.php b/src/ExpressionParser/Prefix/UnaryOperatorExpressionParser.php
--- a/src/ExpressionParser/Prefix/UnaryOperatorExpressionParser.php
+++ b/src/ExpressionParser/Prefix/UnaryOperatorExpressionParser.php
@@ -25,16 +25,31 @@
  */
 final class UnaryOperatorExpressionParser extends AbstractExpressionParser implements PrefixExpressionParserInterface, ExpressionParserDescriptionInterface
 {
+    private string $nodeClass;
+    private string $name;
+    private int $precedence;
+    private ?PrecedenceChange $precedenceChange;
+    private ?string $description;
+    private array $aliases;
+    private ?int $operandPrecedence;
+
     public function __construct(
         /** @var class-string<AbstractUnary> */
-        private string $nodeClass,
-        private string $name,
-        private int $precedence,
-        private ?PrecedenceChange $precedenceChange = null,
-        private ?string $description = null,
-        private array $aliases = [],
-        private ?int $operandPrecedence = null,
+        string $nodeClass,
+        string $name,
+        int $precedence,
+        ?PrecedenceChange $precedenceChange = null,
+        ?string $description = null,
+        array $aliases = [],
+        ?int $operandPrecedence = null
     ) {
+        $this->nodeClass = $nodeClass;
+        $this->name = $name;
+        $this->precedence = $precedence;
+        $this->precedenceChange = $precedenceChange;
+        $this->description = $description;
+        $this->aliases = $aliases;
+        $this->operandPrecedence = $operandPrecedence;
     }
 
     /**
@@ -42,7 +57,8 @@
      */
     public function parse(Parser $parser, Token $token): AbstractExpression
     {
-        return new ($this->nodeClass)($parser->parseExpression($this->operandPrecedence ?? $this->precedence), $token->getLine());
+        $nodeClass = $this->nodeClass;
+        return new $nodeClass($parser->parseExpression($this->operandPrecedence ?? $this->precedence), $token->getLine());
     }
 
     public function getName(): string
diff -u a/src/ExpressionParser.php b/src/ExpressionParser.php
--- a/src/ExpressionParser.php
+++ b/src/ExpressionParser.php
@@ -49,10 +49,16 @@
      */
     public const OPERATOR_RIGHT = 2;
 
+
+    private Parser $parser;
+    private Environment $env;
+
     public function __construct(
-        private Parser $parser,
-        private Environment $env,
+        Parser $parser,
+        Environment $env
     ) {
+        $this->parser = $parser;
+        $this->env = $env;
         trigger_deprecation('twig/twig', '3.21', 'Class "%s" is deprecated, use "Parser::parseExpression()" instead.', __CLASS__);
     }
 
@@ -258,7 +264,7 @@
             $name = null;
             if ($namedArguments && (($token = $stream->nextIf(Token::OPERATOR_TYPE, '=')) || (!$definition && $token = $stream->nextIf(Token::PUNCTUATION_TYPE, ':')))) {
                 if (!$value instanceof ContextVariable) {
-                    throw new SyntaxError(\sprintf('A parameter name must be a string, "%s" given.', $value::class), $token->getLine(), $stream->getSourceContext());
+                    throw new SyntaxError(\sprintf('A parameter name must be a string, "%s" given.', get_class($value)), $token->getLine(), $stream->getSourceContext());
                 }
                 $name = $value->getAttribute('name');
 
diff -u a/src/Extension/AttributeExtension.php b/src/Extension/AttributeExtension.php
--- a/src/Extension/AttributeExtension.php
+++ b/src/Extension/AttributeExtension.php
@@ -29,14 +29,16 @@
     private array $filters;
     private array $functions;
     private array $tests;
+    private string $class;
 
     /**
      * Use a runtime class using PHP attributes to define filters, functions, and tests.
      *
      * @param class-string $class
      */
-    public function __construct(private string $class)
+    public function __construct(string $class)
     {
+        $this->class = $class;
     }
 
     /**
diff -u a/src/Extension/CoreExtension.php b/src/Extension/CoreExtension.php
--- a/src/Extension/CoreExtension.php
+++ b/src/Extension/CoreExtension.php
@@ -26,7 +26,6 @@
 use Twig\ExpressionParser\Infix\IsExpressionParser;
 use Twig\ExpressionParser\Infix\IsNotExpressionParser;
 use Twig\ExpressionParser\Infix\SquareBracketExpressionParser;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\ExpressionParser\PrecedenceChange;
 use Twig\ExpressionParser\Prefix\GroupingExpressionParser;
 use Twig\ExpressionParser\Prefix\LiteralExpressionParser;
@@ -336,13 +335,13 @@
         return [
             // unary operators
             new UnaryOperatorExpressionParser(NotUnary::class, 'not', 50, new PrecedenceChange('twig/twig', '3.15', 70)),
-            new UnaryOperatorExpressionParser(SpreadUnary::class, '...', 512, description: 'Spread operator', operandPrecedence: 0),
+            new UnaryOperatorExpressionParser(SpreadUnary::class, '...', 512, null, 'Spread operator', [], 0),
             new UnaryOperatorExpressionParser(NegUnary::class, '-', 500),
             new UnaryOperatorExpressionParser(PosUnary::class, '+', 500),
 
             // binary operators
-            new BinaryOperatorExpressionParser(ElvisBinary::class, '?:', 5, InfixAssociativity::Right, description: 'Elvis operator (a ?: b)', aliases: ['? :']),
-            new BinaryOperatorExpressionParser(NullCoalesceBinary::class, '??', 300, InfixAssociativity::Right, new PrecedenceChange('twig/twig', '3.15', 5), description: 'Null coalescing operator (a ?? b)'),
+            new BinaryOperatorExpressionParser(ElvisBinary::class, '?:', 5, 2, null, 'Elvis operator (a ?: b)', ['? :']),
+            new BinaryOperatorExpressionParser(NullCoalesceBinary::class, '??', 300, 2, new PrecedenceChange('twig/twig', '3.15', 5), 'Null coalescing operator (a ?? b)'),
             new BinaryOperatorExpressionParser(OrBinary::class, 'or', 10),
             new BinaryOperatorExpressionParser(XorBinary::class, 'xor', 12),
             new BinaryOperatorExpressionParser(AndBinary::class, 'and', 15),
@@ -368,12 +367,12 @@
             new BinaryOperatorExpressionParser(RangeBinary::class, '..', 25),
             new BinaryOperatorExpressionParser(AddBinary::class, '+', 30),
             new BinaryOperatorExpressionParser(SubBinary::class, '-', 30),
-            new BinaryOperatorExpressionParser(ConcatBinary::class, '~', 40, precedenceChange: new PrecedenceChange('twig/twig', '3.15', 27)),
+            new BinaryOperatorExpressionParser(ConcatBinary::class, '~', 40, 1, new PrecedenceChange('twig/twig', '3.15', 27)),
             new BinaryOperatorExpressionParser(MulBinary::class, '*', 60),
             new BinaryOperatorExpressionParser(DivBinary::class, '/', 60),
-            new BinaryOperatorExpressionParser(FloorDivBinary::class, '//', 60, description: 'Floor division'),
+            new BinaryOperatorExpressionParser(FloorDivBinary::class, '//', 60, 1, null, 'Floor division'),
             new BinaryOperatorExpressionParser(ModBinary::class, '%', 60),
-            new BinaryOperatorExpressionParser(PowerBinary::class, '**', 200, InfixAssociativity::Right, description: 'Exponentiation operator'),
+            new BinaryOperatorExpressionParser(PowerBinary::class, '**', 200, 2, null, 'Exponentiation operator'),
 
             // ternary operator
             new ConditionalTernaryExpressionParser(),
@@ -412,7 +411,7 @@
      *
      * @internal
      */
-    public static function cycle($values, $position): mixed
+    public static function cycle($values, $position)/*: mixed*/
     {
         if (!\is_array($values)) {
             if (!$values instanceof \ArrayAccess) {
@@ -960,7 +959,7 @@
      *
      * @internal
      */
-    public static function invoke(\Closure $arrow, ...$arguments): mixed
+    public static function invoke(\Closure $arrow, ...$arguments)/*: mixed*/
     {
         return $arrow(...$arguments);
     }
@@ -1200,18 +1199,25 @@
      *
      * @internal
      */
-    public static function trim($string, $characterMask = null, $side = 'both'): string|\Stringable
+    public static function trim($string, $characterMask = null, $side = 'both')/*: string|\Stringable*/
     {
         if (null === $characterMask) {
             $characterMask = self::DEFAULT_TRIM_CHARS;
         }
 
-        $trimmed = match ($side) {
-            'both' => trim($string ?? '', $characterMask),
-            'left' => ltrim($string ?? '', $characterMask),
-            'right' => rtrim($string ?? '', $characterMask),
-            default => throw new RuntimeError('Trimming side must be "left", "right" or "both".'),
-        };
+        switch ($side) {
+            case 'both':
+                $trimmed = trim($string ?? '', $characterMask);
+                break;
+            case 'left':
+                $trimmed = ltrim($string ?? '', $characterMask);
+                break;
+            case 'right':
+                $trimmed = rtrim($string ?? '', $characterMask);
+                break;
+            default:
+                throw new RuntimeError('Trimming side must be "left", "right" or "both".');
+        }
 
         // trimming a safe string with the default character mask always returns a safe string (independently of the context)
         return $string instanceof Markup && self::DEFAULT_TRIM_CHARS === $characterMask ? new Markup($trimmed, $string->getCharset()) : $trimmed;
@@ -1580,7 +1586,7 @@
      *
      * @internal
      */
-    public static function enum(string $enum): \UnitEnum
+    public static function enum(string $enum)/*: \UnitEnum*/
     {
         if (!enum_exists($enum)) {
             throw new RuntimeError(\sprintf('"%s" is not an enum.', $enum));
@@ -1610,10 +1616,10 @@
     {
         if (null !== $object) {
             if ('class' === $constant) {
-                return $checkDefined ? true : $object::class;
+                return $checkDefined ? true : get_class($object);
             }
 
-            $constant = $object::class.'::'.$constant;
+            $constant = get_class($object).'::'.$constant;
         }
 
         if (!\defined($constant)) {
@@ -1690,7 +1696,7 @@
         if (Template::METHOD_CALL !== $type) {
             $arrayItem = \is_bool($item) || \is_float($item) ? (int) $item : $item;
 
-            if ($sandboxed && $object instanceof \ArrayAccess && !\in_array($object::class, self::ARRAY_LIKE_CLASSES, true)) {
+            if ($sandboxed && $object instanceof \ArrayAccess && !\in_array(get_class($object), self::ARRAY_LIKE_CLASSES, true)) {
                 try {
                     $env->getExtension(SandboxExtension::class)->checkPropertyAllowed($object, $arrayItem, $lineno, $source);
                 } catch (SecurityNotAllowedPropertyError $propertyNotAllowedError) {
@@ -1702,11 +1708,10 @@
                 }
             }
 
-            if (match (true) {
-                \is_array($object) => \array_key_exists($arrayItem = (string) $arrayItem, $object),
-                $object instanceof \ArrayAccess => $object->offsetExists($arrayItem),
-                default => false,
-            }) {
+            if (
+                (\is_array($object) && \array_key_exists($arrayItem = (string) $arrayItem, $object))
+                || ($object instanceof \ArrayAccess && $object->offsetExists($arrayItem))
+            ) {
                 if ($isDefinedTest) {
                     return true;
                 }
@@ -1792,7 +1797,7 @@
             static $propertyCheckers = [];
 
             if (isset($object->$item)
-                || ($propertyCheckers[$object::class][$item] ??= self::getPropertyChecker($object::class, $item))($object, $item)
+                || ($propertyCheckers[get_class($object)][$item] ??= self::getPropertyChecker(get_class($object), $item))($object, $item)
             ) {
                 if ($isDefinedTest) {
                     return true;
@@ -1809,12 +1814,12 @@
                 return ((array) $object)[$item];
             }
 
-            if (\defined($object::class.'::'.$item)) {
+            if (\defined(get_class($object).'::'.$item)) {
                 if ($isDefinedTest) {
                     return true;
                 }
 
-                return \constant($object::class.'::'.$item);
+                return \constant(get_class($object).'::'.$item);
             }
         }
 
@@ -1822,7 +1827,7 @@
 
         static $cache = [];
 
-        $class = $object::class;
+        $class = get_class($object);
 
         // object method
         // precedence: getXxx() > isXxx() > hasXxx()
diff -u a/src/Extension/OptimizerExtension.php b/src/Extension/OptimizerExtension.php
--- a/src/Extension/OptimizerExtension.php
+++ b/src/Extension/OptimizerExtension.php
@@ -15,9 +15,12 @@
 
 final class OptimizerExtension extends AbstractExtension
 {
+    private int $optimizers;
+
     public function __construct(
-        private int $optimizers = -1,
+        int $optimizers = -1
     ) {
+        $this->optimizers = $optimizers;
     }
 
     public function getNodeVisitors(): array
diff -u a/src/Extension/YieldNotReadyExtension.php b/src/Extension/YieldNotReadyExtension.php
--- a/src/Extension/YieldNotReadyExtension.php
+++ b/src/Extension/YieldNotReadyExtension.php
@@ -18,9 +18,12 @@
  */
 final class YieldNotReadyExtension extends AbstractExtension
 {
+    private bool $useYield;
+
     public function __construct(
-        private bool $useYield,
+        bool $useYield
     ) {
+        $this->useYield = $useYield;
     }
 
     public function getNodeVisitors(): array
diff -u a/src/ExtensionSet.php b/src/ExtensionSet.php
--- a/src/ExtensionSet.php
+++ b/src/ExtensionSet.php
@@ -14,7 +14,6 @@
 use Twig\Error\RuntimeError;
 use Twig\ExpressionParser\ExpressionParsers;
 use Twig\ExpressionParser\Infix\BinaryOperatorExpressionParser;
-use Twig\ExpressionParser\InfixAssociativity;
 use Twig\ExpressionParser\InfixExpressionParserInterface;
 use Twig\ExpressionParser\PrecedenceChange;
 use Twig\ExpressionParser\Prefix\UnaryOperatorExpressionParser;
@@ -152,7 +151,7 @@
         if ($extension instanceof AttributeExtension) {
             $class = $extension->getClass();
         } else {
-            $class = $extension::class;
+            $class = \get_class($extension);
         }
 
         if ($this->initialized) {
@@ -509,11 +508,11 @@
 
         $operators = $extension->getOperators();
         if (!\is_array($operators)) {
-            throw new \InvalidArgumentException(\sprintf('"%s::getOperators()" must return an array with operators, got "%s".', $extension::class, get_debug_type($operators).(\is_resource($operators) ? '' : '#'.$operators)));
+            throw new \InvalidArgumentException(\sprintf('"%s::getOperators()" must return an array with operators, got "%s".', get_class($extension), get_debug_type($operators).(\is_resource($operators) ? '' : '#'.$operators)));
         }
 
         if (2 !== \count($operators)) {
-            throw new \InvalidArgumentException(\sprintf('"%s::getOperators()" must return an array of 2 elements, got %d.', $extension::class, \count($operators)));
+            throw new \InvalidArgumentException(\sprintf('"%s::getOperators()" must return an array of 2 elements, got %d.', get_class($extension), \count($operators)));
         }
 
         $expressionParsers = [];
@@ -521,12 +520,6 @@
             $expressionParsers[] = new UnaryOperatorExpressionParser($op['class'], $operator, $op['precedence'], $op['precedence_change'] ?? null, '', $op['aliases'] ?? []);
         }
         foreach ($operators[1] as $operator => $op) {
-            $op['associativity'] = match ($op['associativity']) {
-                1 => InfixAssociativity::Left,
-                2 => InfixAssociativity::Right,
-                default => throw new \InvalidArgumentException(\sprintf('Invalid associativity "%s" for operator "%s".', $op['associativity'], $operator)),
-            };
-
             if (isset($op['callable'])) {
                 $expressionParsers[] = $this->convertInfixExpressionParser($op['class'], $operator, $op['precedence'], $op['associativity'], $op['precedence_change'] ?? null, $op['aliases'] ?? [], $op['callable']);
             } else {
@@ -535,26 +528,29 @@
         }
 
         if (\count($expressionParsers)) {
-            trigger_deprecation('twig/twig', '3.21', \sprintf('Extension "%s" uses the old signature for "getOperators()", please implement "getExpressionParsers()" instead.', $extension::class));
+            trigger_deprecation('twig/twig', '3.21', \sprintf('Extension "%s" uses the old signature for "getOperators()", please implement "getExpressionParsers()" instead.', get_class($extension)));
 
             $this->expressionParsers->add($expressionParsers);
         }
     }
 
-    private function convertInfixExpressionParser(string $nodeClass, string $operator, int $precedence, InfixAssociativity $associativity, ?PrecedenceChange $precedenceChange, array $aliases, callable $callable): InfixExpressionParserInterface
+    private function convertInfixExpressionParser(string $nodeClass, string $operator, int $precedence, int $associativity, ?PrecedenceChange $precedenceChange, array $aliases, callable $callable): InfixExpressionParserInterface
     {
         trigger_deprecation('twig/twig', '3.21', \sprintf('Using a non-ExpressionParserInterface object to define the "%s" binary operator is deprecated.', $operator));
 
         return new class($nodeClass, $operator, $precedence, $associativity, $precedenceChange, $aliases, $callable) extends BinaryOperatorExpressionParser {
+            private $callable = null;
+
             public function __construct(
                 string $nodeClass,
                 string $operator,
                 int $precedence,
-                InfixAssociativity $associativity = InfixAssociativity::Left,
+                int $associativity = 1,
                 ?PrecedenceChange $precedenceChange = null,
                 array $aliases = [],
-                private $callable = null,
+                $callable = null
             ) {
+                $this->callable = $callable;
                 parent::__construct($nodeClass, $operator, $precedence, $associativity, $precedenceChange, $aliases);
             }
 
diff -u a/src/Loader/ArrayLoader.php b/src/Loader/ArrayLoader.php
--- a/src/Loader/ArrayLoader.php
+++ b/src/Loader/ArrayLoader.php
@@ -28,12 +28,15 @@
  */
 final class ArrayLoader implements LoaderInterface
 {
+    private array $templates;
+
     /**
      * @param array $templates An array of templates (keys are the names, and values are the source code)
      */
     public function __construct(
-        private array $templates = [],
+        array $templates = []
     ) {
+        $this->templates = $templates;
     }
 
     public function setTemplate(string $name, string $template): void
diff -u a/src/Loader/ChainLoader.php b/src/Loader/ChainLoader.php
--- a/src/Loader/ChainLoader.php
+++ b/src/Loader/ChainLoader.php
@@ -25,13 +25,15 @@
      * @var array<string, bool>
      */
     private $hasSourceCache = [];
+    private iterable $loaders;
 
     /**
      * @param iterable<LoaderInterface> $loaders
      */
     public function __construct(
-        private iterable $loaders = [],
+        iterable $loaders = []
     ) {
+        $this->loaders = $loaders;
     }
 
     public function addLoader(LoaderInterface $loader): void
@@ -104,7 +106,7 @@
             try {
                 return $loader->getCacheKey($name);
             } catch (LoaderError $e) {
-                $exceptions[] = $loader::class.': '.$e->getMessage();
+                $exceptions[] = get_class($loader).': '.$e->getMessage();
             }
         }
 
@@ -123,7 +125,7 @@
             try {
                 return $loader->isFresh($name, $time);
             } catch (LoaderError $e) {
-                $exceptions[] = $loader::class.': '.$e->getMessage();
+                $exceptions[] = get_class($loader).': '.$e->getMessage();
             }
         }
 
diff -u a/src/Node/Expression/Binary/AbstractBinary.php b/src/Node/Expression/Binary/AbstractBinary.php
--- a/src/Node/Expression/Binary/AbstractBinary.php
+++ b/src/Node/Expression/Binary/AbstractBinary.php
@@ -25,10 +25,10 @@
     public function __construct(Node $left, Node $right, int $lineno)
     {
         if (!$left instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "left" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $left::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "left" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($left));
         }
         if (!$right instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "right" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $right::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "right" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($right));
         }
 
         parent::__construct(['left' => $left, 'right' => $right], [], $lineno);
diff -u a/src/Node/Expression/Binary/MatchesBinary.php b/src/Node/Expression/Binary/MatchesBinary.php
--- a/src/Node/Expression/Binary/MatchesBinary.php
+++ b/src/Node/Expression/Binary/MatchesBinary.php
@@ -24,15 +24,15 @@
     public function __construct(Node $left, Node $right, int $lineno)
     {
         if (!$left instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.24', 'Passing a "%s" instance to "%s()" first argument is deprecated, pass an "AbstractExpression" instance instead.', $left::class, __METHOD__);
+            trigger_deprecation('twig/twig', '3.24', 'Passing a "%s" instance to "%s()" first argument is deprecated, pass an "AbstractExpression" instance instead.', get_class($left), __METHOD__);
         }
         if (!$right instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.24', 'Passing a "%s" instance to "%s()" second argument is deprecated, pass an "AbstractExpression" instance instead.', $right::class, __METHOD__);
+            trigger_deprecation('twig/twig', '3.24', 'Passing a "%s" instance to "%s()" second argument is deprecated, pass an "AbstractExpression" instance instead.', get_class($right), __METHOD__);
         }
 
         if ($right instanceof ConstantExpression) {
             $regexp = $right->getAttribute('value');
-            set_error_handler(static fn ($t, $m) => throw new SyntaxError(\sprintf('Regexp "%s" passed to "matches" is not valid: %s.', $regexp, substr($m, 14)), $lineno));
+            set_error_handler(static function ($t, $m) use ($regexp, $lineno) { throw new SyntaxError(\sprintf('Regexp "%s" passed to "matches" is not valid: %s.', $regexp, substr($m, 14)), $lineno); });
             try {
                 preg_match($regexp, '');
             } finally {
diff -u a/src/Node/Expression/Binary/ObjectDestructuringSetBinary.php b/src/Node/Expression/Binary/ObjectDestructuringSetBinary.php
--- a/src/Node/Expression/Binary/ObjectDestructuringSetBinary.php
+++ b/src/Node/Expression/Binary/ObjectDestructuringSetBinary.php
@@ -38,7 +38,7 @@
         }
         foreach ($left->getKeyValuePairs() as $pair) {
             if (!$pair['value'] instanceof ContextVariable) {
-                throw new SyntaxError(\sprintf('Cannot assign to "%s", only variables can be assigned in object/mapping destructuring.', $pair['value']::class), $lineno);
+                throw new SyntaxError(\sprintf('Cannot assign to "%s", only variables can be assigned in object/mapping destructuring.', get_class($pair['value'])), $lineno);
             }
 
             $this->mappings[] = [
diff -u a/src/Node/Expression/Binary/SequenceDestructuringSetBinary.php b/src/Node/Expression/Binary/SequenceDestructuringSetBinary.php
--- a/src/Node/Expression/Binary/SequenceDestructuringSetBinary.php
+++ b/src/Node/Expression/Binary/SequenceDestructuringSetBinary.php
@@ -38,7 +38,7 @@
             } elseif ($pair['value'] instanceof ContextVariable) {
                 $this->variables[] = $pair['value']->getAttribute('name');
             } else {
-                throw new SyntaxError(\sprintf('Cannot assign to "%s", only variables can be assigned in sequence destructuring.', $pair['value']::class), $lineno);
+                throw new SyntaxError(\sprintf('Cannot assign to "%s", only variables can be assigned in sequence destructuring.', get_class($pair['value'])), $lineno);
             }
         }
 
diff -u a/src/Node/Expression/BlockReferenceExpression.php b/src/Node/Expression/BlockReferenceExpression.php
--- a/src/Node/Expression/BlockReferenceExpression.php
+++ b/src/Node/Expression/BlockReferenceExpression.php
@@ -32,7 +32,7 @@
     public function __construct(Node $name, ?Node $template, int $lineno)
     {
         if (!$name instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $name::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($name));
         }
 
         $nodes = ['name' => $name];
diff -u a/src/Node/Expression/CallExpression.php b/src/Node/Expression/CallExpression.php
--- a/src/Node/Expression/CallExpression.php
+++ b/src/Node/Expression/CallExpression.php
@@ -342,7 +342,7 @@
             return $twigCallable->needsIsSandboxed();
         }
 
-        trigger_deprecation('twig/twig', '3.25', 'Not implementing the "needsIsSandboxed()" method in "%s" is deprecated. This method will be part of the "%s" interface in 4.0.', $twigCallable::class, TwigCallableInterface::class);
+        trigger_deprecation('twig/twig', '3.25', 'Not implementing the "needsIsSandboxed()" method in "%s" is deprecated. This method will be part of the "%s" interface in 4.0.', get_class($twigCallable), TwigCallableInterface::class);
 
         return false;
     }
@@ -356,38 +356,47 @@
     {
         $current = $this->getAttribute('twig_callable');
 
-        $this->setAttribute('twig_callable', match ($this->getAttribute('type')) {
-            'test' => (new TwigTest(
+        $value = '';
+        switch($this->getAttribute('type')) {
+            case 'test':
+                $value = (new TwigTest(
+                    $this->getAttribute('name'),
+                    $this->hasAttribute('callable') ? $this->getAttribute('callable') : $current->getCallable(),
+                    [
+                        'needs_is_sandboxed' => $this->hasAttribute('needs_is_sandboxed') ? $this->getAttribute('needs_is_sandboxed') : self::needsIsSandboxed($current),
+                        'is_variadic' => $this->hasAttribute('is_variadic') ? $this->getAttribute('is_variadic') : $current->isVariadic(),
+                    ],
+                ))->withDynamicArguments($this->getAttribute('name'), $this->hasAttribute('dynamic_name') ? $this->getAttribute('dynamic_name') : $current->getDynamicName(), $this->hasAttribute('arguments') ? $this->getAttribute('arguments') : $current->getArguments());
+                break;
+            case 'function':
+                $value = (new TwigFunction(
+                    $this->hasAttribute('name') ? $this->getAttribute('name') : $current->getName(),
+                    $this->hasAttribute('callable') ? $this->getAttribute('callable') : $current->getCallable(),
+                    [
+                        'needs_environment' => $this->hasAttribute('needs_environment') ? $this->getAttribute('needs_environment') : $current->needsEnvironment(),
+                        'needs_context' => $this->hasAttribute('needs_context') ? $this->getAttribute('needs_context') : $current->needsContext(),
+                        'needs_charset' => $this->hasAttribute('needs_charset') ? $this->getAttribute('needs_charset') : $current->needsCharset(),
+                        'needs_is_sandboxed' => $this->hasAttribute('needs_is_sandboxed') ? $this->getAttribute('needs_is_sandboxed') : self::needsIsSandboxed($current),
+                        'is_variadic' => $this->hasAttribute('is_variadic') ? $this->getAttribute('is_variadic') : $current->isVariadic(),
+                    ],
+                ))->withDynamicArguments($this->getAttribute('name'), $this->hasAttribute('dynamic_name') ? $this->getAttribute('dynamic_name') : $current->getDynamicName(), $this->hasAttribute('arguments') ? $this->getAttribute('arguments') : $current->getArguments());
+                break;
+            case 'filter':
+                $value = (new TwigFilter(
                 $this->getAttribute('name'),
                 $this->hasAttribute('callable') ? $this->getAttribute('callable') : $current->getCallable(),
-                [
-                    'needs_is_sandboxed' => $this->hasAttribute('needs_is_sandboxed') ? $this->getAttribute('needs_is_sandboxed') : self::needsIsSandboxed($current),
-                    'is_variadic' => $this->hasAttribute('is_variadic') ? $this->getAttribute('is_variadic') : $current->isVariadic(),
-                ],
-            ))->withDynamicArguments($this->getAttribute('name'), $this->hasAttribute('dynamic_name') ? $this->getAttribute('dynamic_name') : $current->getDynamicName(), $this->hasAttribute('arguments') ? $this->getAttribute('arguments') : $current->getArguments()),
-            'function' => (new TwigFunction(
-                $this->hasAttribute('name') ? $this->getAttribute('name') : $current->getName(),
-                $this->hasAttribute('callable') ? $this->getAttribute('callable') : $current->getCallable(),
-                [
-                    'needs_environment' => $this->hasAttribute('needs_environment') ? $this->getAttribute('needs_environment') : $current->needsEnvironment(),
-                    'needs_context' => $this->hasAttribute('needs_context') ? $this->getAttribute('needs_context') : $current->needsContext(),
-                    'needs_charset' => $this->hasAttribute('needs_charset') ? $this->getAttribute('needs_charset') : $current->needsCharset(),
-                    'needs_is_sandboxed' => $this->hasAttribute('needs_is_sandboxed') ? $this->getAttribute('needs_is_sandboxed') : self::needsIsSandboxed($current),
-                    'is_variadic' => $this->hasAttribute('is_variadic') ? $this->getAttribute('is_variadic') : $current->isVariadic(),
-                ],
-            ))->withDynamicArguments($this->getAttribute('name'), $this->hasAttribute('dynamic_name') ? $this->getAttribute('dynamic_name') : $current->getDynamicName(), $this->hasAttribute('arguments') ? $this->getAttribute('arguments') : $current->getArguments()),
-            'filter' => (new TwigFilter(
-                $this->getAttribute('name'),
-                $this->hasAttribute('callable') ? $this->getAttribute('callable') : $current->getCallable(),
-                [
-                    'needs_environment' => $this->hasAttribute('needs_environment') ? $this->getAttribute('needs_environment') : $current->needsEnvironment(),
-                    'needs_context' => $this->hasAttribute('needs_context') ? $this->getAttribute('needs_context') : $current->needsContext(),
-                    'needs_charset' => $this->hasAttribute('needs_charset') ? $this->getAttribute('needs_charset') : $current->needsCharset(),
-                    'needs_is_sandboxed' => $this->hasAttribute('needs_is_sandboxed') ? $this->getAttribute('needs_is_sandboxed') : self::needsIsSandboxed($current),
-                    'is_variadic' => $this->hasAttribute('is_variadic') ? $this->getAttribute('is_variadic') : $current->isVariadic(),
-                ],
-            ))->withDynamicArguments($this->getAttribute('name'), $this->hasAttribute('dynamic_name') ? $this->getAttribute('dynamic_name') : $current->getDynamicName(), $this->hasAttribute('arguments') ? $this->getAttribute('arguments') : $current->getArguments()),
-        });
+                    [
+                        'needs_environment' => $this->hasAttribute('needs_environment') ? $this->getAttribute('needs_environment') : $current->needsEnvironment(),
+                        'needs_context' => $this->hasAttribute('needs_context') ? $this->getAttribute('needs_context') : $current->needsContext(),
+                        'needs_charset' => $this->hasAttribute('needs_charset') ? $this->getAttribute('needs_charset') : $current->needsCharset(),
+                        'needs_is_sandboxed' => $this->hasAttribute('needs_is_sandboxed') ? $this->getAttribute('needs_is_sandboxed') : self::needsIsSandboxed($current),
+                        'is_variadic' => $this->hasAttribute('is_variadic') ? $this->getAttribute('is_variadic') : $current->isVariadic(),
+                    ],
+                ))->withDynamicArguments($this->getAttribute('name'), $this->hasAttribute('dynamic_name') ? $this->getAttribute('dynamic_name') : $current->getDynamicName(), $this->hasAttribute('arguments') ? $this->getAttribute('arguments') : $current->getArguments());
+                break;
+        }
+
+        $this->setAttribute('twig_callable', $value);
 
         return $this->getAttribute('twig_callable');
     }
diff -u a/src/Node/Expression/Filter/DefaultFilter.php b/src/Node/Expression/Filter/DefaultFilter.php
--- a/src/Node/Expression/Filter/DefaultFilter.php
+++ b/src/Node/Expression/Filter/DefaultFilter.php
@@ -39,10 +39,10 @@
      * @param AbstractExpression $node
      */
     #[FirstClassTwigCallableReady]
-    public function __construct(Node $node, TwigFilter|ConstantExpression $filter, Node $arguments, int $lineno)
+    public function __construct(Node $node, /*TwigFilter|ConstantExpression*/ $filter, Node $arguments, int $lineno)
     {
         if (!$node instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $node::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($node));
         }
 
         if ($filter instanceof TwigFilter) {
diff -u a/src/Node/Expression/Filter/RawFilter.php b/src/Node/Expression/Filter/RawFilter.php
--- a/src/Node/Expression/Filter/RawFilter.php
+++ b/src/Node/Expression/Filter/RawFilter.php
@@ -29,10 +29,10 @@
      * @param AbstractExpression $node
      */
     #[FirstClassTwigCallableReady]
-    public function __construct(Node $node, TwigFilter|ConstantExpression|null $filter = null, ?Node $arguments = null, int $lineno = 0)
+    public function __construct(Node $node, /*TwigFilter|ConstantExpression|null*/ $filter = null, ?Node $arguments = null, int $lineno = 0)
     {
         if (!$node instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $node::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($node));
         }
 
         parent::__construct($node, $filter ?: new TwigFilter('raw', null, ['is_safe' => ['all']]), $arguments ?: new EmptyNode(), $lineno ?: $node->getTemplateLine());
diff -u a/src/Node/Expression/FilterExpression.php b/src/Node/Expression/FilterExpression.php
--- a/src/Node/Expression/FilterExpression.php
+++ b/src/Node/Expression/FilterExpression.php
@@ -25,10 +25,10 @@
      * @param AbstractExpression $node
      */
     #[FirstClassTwigCallableReady]
-    public function __construct(Node $node, TwigFilter|ConstantExpression $filter, Node $arguments, int $lineno)
+    public function __construct(Node $node, /*TwigFilter|ConstantExpression*/ $filter, Node $arguments, int $lineno)
     {
         if (!$node instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $node::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($node));
         }
 
         if ($filter instanceof TwigFilter) {
diff -u a/src/Node/Expression/FunctionExpression.php b/src/Node/Expression/FunctionExpression.php
--- a/src/Node/Expression/FunctionExpression.php
+++ b/src/Node/Expression/FunctionExpression.php
@@ -24,7 +24,7 @@
     use SupportDefinedTestTrait;
 
     #[FirstClassTwigCallableReady]
-    public function __construct(TwigFunction|string $function, Node $arguments, int $lineno)
+    public function __construct(/*TwigFunction|string*/ $function, Node $arguments, int $lineno)
     {
         if ($function instanceof TwigFunction) {
             $name = $function->getName();
diff -u a/src/Node/Expression/GetAttrExpression.php b/src/Node/Expression/GetAttrExpression.php
--- a/src/Node/Expression/GetAttrExpression.php
+++ b/src/Node/Expression/GetAttrExpression.php
@@ -34,7 +34,7 @@
         }
 
         if ($arguments && !$arguments instanceof ArrayExpression && !$arguments instanceof ContextVariable) {
-            trigger_deprecation('twig/twig', '3.15', \sprintf('Not passing a "%s" instance as the "arguments" argument of the "%s" constructor is deprecated ("%s" given).', ArrayExpression::class, static::class, $arguments::class));
+            trigger_deprecation('twig/twig', '3.15', \sprintf('Not passing a "%s" instance as the "arguments" argument of the "%s" constructor is deprecated ("%s" given).', ArrayExpression::class, static::class, get_class($arguments)));
         }
 
         parent::__construct($nodes, ['type' => $type, 'ignore_strict_check' => false, 'optimizable' => !$nullSafe, 'null_safe' => $nullSafe, 'is_short_circuited' => false, 'var_name' => null], $lineno);
diff -u a/src/Node/Expression/NullCoalesceExpression.php b/src/Node/Expression/NullCoalesceExpression.php
--- a/src/Node/Expression/NullCoalesceExpression.php
+++ b/src/Node/Expression/NullCoalesceExpression.php
@@ -33,10 +33,10 @@
         trigger_deprecation('twig/twig', '3.17', \sprintf('"%s" is deprecated; use "%s" instead.', __CLASS__, NullCoalesceBinary::class));
 
         if (!$left instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "left" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $left::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "left" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($left));
         }
         if (!$right instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "right" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $right::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "right" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($right));
         }
 
         $test = new DefinedTest(clone $left, new TwigTest('defined'), new EmptyNode(), $left->getTemplateLine());
diff -u a/src/Node/Expression/TempNameExpression.php b/src/Node/Expression/TempNameExpression.php
--- a/src/Node/Expression/TempNameExpression.php
+++ b/src/Node/Expression/TempNameExpression.php
@@ -18,7 +18,7 @@
 {
     public const RESERVED_NAMES = ['varargs', 'context', 'macros', 'blocks', 'this'];
 
-    public function __construct(string|int|null $name, int $lineno)
+    public function __construct(/*string|int|null*/ $name, int $lineno)
     {
         // All names supported by ExpressionParser::parsePrimaryExpression() should be excluded
         if ($name && \in_array(strtolower($name), ['true', 'false', 'none', 'null'], true)) {
diff -u a/src/Node/Expression/Test/DefinedTest.php b/src/Node/Expression/Test/DefinedTest.php
--- a/src/Node/Expression/Test/DefinedTest.php
+++ b/src/Node/Expression/Test/DefinedTest.php
@@ -36,10 +36,10 @@
      * @param AbstractExpression $node
      */
     #[FirstClassTwigCallableReady]
-    public function __construct(Node $node, TwigTest|string $name, ?Node $arguments, int $lineno)
+    public function __construct(Node $node, /*TwigTest|string*/ $name, ?Node $arguments, int $lineno)
     {
         if (!$node instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $node::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($node));
         }
 
         if (!$node instanceof SupportDefinedTestInterface) {
diff -u a/src/Node/Expression/TestExpression.php b/src/Node/Expression/TestExpression.php
--- a/src/Node/Expression/TestExpression.php
+++ b/src/Node/Expression/TestExpression.php
@@ -24,10 +24,10 @@
     /**
      * @param AbstractExpression $node
      */
-    public function __construct(Node $node, string|TwigTest $test, ?Node $arguments, int $lineno)
+    public function __construct(Node $node, /*string|TwigTest*/ $test, ?Node $arguments, int $lineno)
     {
         if (!$node instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $node::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance to the "node" argument of "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($node));
         }
 
         $nodes = ['node' => $node];
diff -u a/src/Node/Expression/Unary/AbstractUnary.php b/src/Node/Expression/Unary/AbstractUnary.php
--- a/src/Node/Expression/Unary/AbstractUnary.php
+++ b/src/Node/Expression/Unary/AbstractUnary.php
@@ -24,7 +24,7 @@
     public function __construct(Node $node, int $lineno)
     {
         if (!$node instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance argument to "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, $node::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance argument to "%s" is deprecated ("%s" given).', AbstractExpression::class, static::class, get_class($node));
         }
 
         parent::__construct(['node' => $node], ['with_parentheses' => false], $lineno);
diff -u a/src/Node/ImportNode.php b/src/Node/ImportNode.php
--- a/src/Node/ImportNode.php
+++ b/src/Node/ImportNode.php
@@ -25,14 +25,14 @@
 #[YieldReady]
 class ImportNode extends Node implements CoercesChildrenToStringInterface
 {
-    public function __construct(AbstractExpression $expr, AbstractExpression|AssignTemplateVariable $var, int $lineno)
+    public function __construct(AbstractExpression $expr, /*AbstractExpression|AssignTemplateVariable*/ $var, int $lineno)
     {
         if (\func_num_args() > 3) {
             trigger_deprecation('twig/twig', '3.15', \sprintf('Passing more than 3 arguments to "%s()" is deprecated.', __METHOD__));
         }
 
         if (!$var instanceof AssignTemplateVariable) {
-            trigger_deprecation('twig/twig', '3.15', \sprintf('Passing a "%s" instance as the second argument of "%s" is deprecated, pass a "%s" instead.', $var::class, __CLASS__, AssignTemplateVariable::class));
+            trigger_deprecation('twig/twig', '3.15', \sprintf('Passing a "%s" instance as the second argument of "%s" is deprecated, pass a "%s" instead.', get_class($var), __CLASS__, AssignTemplateVariable::class));
 
             $var = new AssignTemplateVariable($var->getAttribute('name'), $lineno);
         }
diff -u a/src/Node/MacroNode.php b/src/Node/MacroNode.php
--- a/src/Node/MacroNode.php
+++ b/src/Node/MacroNode.php
@@ -34,11 +34,11 @@
     public function __construct(string $name, Node $body, Node $arguments, int $lineno)
     {
         if (!$body instanceof BodyNode) {
-            trigger_deprecation('twig/twig', '3.12', \sprintf('Not passing a "%s" instance as the "body" argument of the "%s" constructor is deprecated ("%s" given).', BodyNode::class, static::class, $body::class));
+            trigger_deprecation('twig/twig', '3.12', \sprintf('Not passing a "%s" instance as the "body" argument of the "%s" constructor is deprecated ("%s" given).', BodyNode::class, static::class, get_class($body)));
         }
 
         if (!$arguments instanceof ArrayExpression) {
-            trigger_deprecation('twig/twig', '3.15', \sprintf('Not passing a "%s" instance as the "arguments" argument of the "%s" constructor is deprecated ("%s" given).', ArrayExpression::class, static::class, $arguments::class));
+            trigger_deprecation('twig/twig', '3.15', \sprintf('Not passing a "%s" instance as the "arguments" argument of the "%s" constructor is deprecated ("%s" given).', ArrayExpression::class, static::class, get_class($arguments)));
 
             $args = new ArrayExpression([], $arguments->getTemplateLine());
             foreach ($arguments as $n => $default) {
diff -u a/src/Node/SetNode.php b/src/Node/SetNode.php
--- a/src/Node/SetNode.php
+++ b/src/Node/SetNode.php
@@ -34,7 +34,7 @@
         if ($capture) {
             $safe = true;
             // Node::class === get_class($values) should be removed in Twig 4.0
-            if (($values instanceof Nodes || Node::class === $values::class) && !\count($values)) {
+            if (($values instanceof Nodes || Node::class === get_class($values)) && !\count($values)) {
                 $values = new ConstantExpression('', $values->getTemplateLine());
                 $capture = false;
             } elseif ($values instanceof TextNode) {
diff -u a/src/NodeVisitor/EscaperNodeVisitor.php b/src/NodeVisitor/EscaperNodeVisitor.php
--- a/src/NodeVisitor/EscaperNodeVisitor.php
+++ b/src/NodeVisitor/EscaperNodeVisitor.php
@@ -160,7 +160,7 @@
     /**
      * @return string|false
      */
-    private function needEscaping(): string|bool
+    private function needEscaping()/*: string|bool*/
     {
         if (\count($this->statusStack)) {
             return $this->statusStack[\count($this->statusStack) - 1];
diff -u a/src/NodeVisitor/OptimizerNodeVisitor.php b/src/NodeVisitor/OptimizerNodeVisitor.php
--- a/src/NodeVisitor/OptimizerNodeVisitor.php
+++ b/src/NodeVisitor/OptimizerNodeVisitor.php
@@ -48,12 +48,16 @@
     private $loops = [];
     private $loopsTargets = [];
 
+    private int $optimizers;
+
     /**
      * @param int $optimizers The optimizer mode
      */
     public function __construct(
-        private int $optimizers = -1,
+        int $optimizers = -1
     ) {
+        $this->optimizers = $optimizers;
+
         if ($optimizers > (self::OPTIMIZE_FOR | self::OPTIMIZE_RAW_FILTER | self::OPTIMIZE_TEXT_NODES)) {
             throw new \InvalidArgumentException(\sprintf('Optimizer mode "%s" is not valid.', $optimizers));
         }
diff -u a/src/NodeVisitor/SafeAnalysisNodeVisitor.php b/src/NodeVisitor/SafeAnalysisNodeVisitor.php
--- a/src/NodeVisitor/SafeAnalysisNodeVisitor.php
+++ b/src/NodeVisitor/SafeAnalysisNodeVisitor.php
@@ -122,7 +122,7 @@
             if ($filter) {
                 $safe = $filter->getSafe($node->getNode('arguments'));
                 if (null === $safe) {
-                    trigger_deprecation('twig/twig', '3.16', 'The "%s::getSafe()" method should not return "null" anymore, return "[]" instead.', $filter::class);
+                    trigger_deprecation('twig/twig', '3.16', 'The "%s::getSafe()" method should not return "null" anymore, return "[]" instead.', get_class($filter));
                     $safe = [];
                 }
 
@@ -143,7 +143,7 @@
             if ($function) {
                 $safe = $function->getSafe($node->getNode('arguments'));
                 if (null === $safe) {
-                    trigger_deprecation('twig/twig', '3.16', 'The "%s::getSafe()" method should not return "null" anymore, return "[]" instead.', $function::class);
+                    trigger_deprecation('twig/twig', '3.16', 'The "%s::getSafe()" method should not return "null" anymore, return "[]" instead.', get_class($function));
                     $safe = [];
                 }
                 $this->setSafe($node, $safe);
diff -u a/src/NodeVisitor/YieldNotReadyNodeVisitor.php b/src/NodeVisitor/YieldNotReadyNodeVisitor.php
--- a/src/NodeVisitor/YieldNotReadyNodeVisitor.php
+++ b/src/NodeVisitor/YieldNotReadyNodeVisitor.php
@@ -22,15 +22,17 @@
 final class YieldNotReadyNodeVisitor implements NodeVisitorInterface
 {
     private $yieldReadyNodes = [];
+    private bool $useYield;
 
     public function __construct(
-        private bool $useYield,
+        bool $useYield
     ) {
+        $this->useYield = $useYield;
     }
 
     public function enterNode(Node $node, Environment $env): Node
     {
-        $class = $node::class;
+        $class = get_class($node);
 
         if ($node instanceof AbstractExpression || isset($this->yieldReadyNodes[$class])) {
             return $node;
diff -u a/src/OperatorPrecedenceChange.php b/src/OperatorPrecedenceChange.php
--- a/src/OperatorPrecedenceChange.php
+++ b/src/OperatorPrecedenceChange.php
@@ -23,9 +23,9 @@
 class OperatorPrecedenceChange extends PrecedenceChange
 {
     public function __construct(
-        private string $package,
-        private string $version,
-        private int $newPrecedence,
+        string $package,
+        string $version,
+        int $newPrecedence
     ) {
         trigger_deprecation('twig/twig', '3.21', 'The "%s" class is deprecated since Twig 3.21. Use "%s" instead.', self::class, PrecedenceChange::class);
 
diff -u a/src/Parser.php b/src/Parser.php
--- a/src/Parser.php
+++ b/src/Parser.php
@@ -15,7 +15,6 @@
 use Twig\Error\SyntaxError;
 use Twig\ExpressionParser\ExpressionParserInterface;
 use Twig\ExpressionParser\ExpressionParsers;
-use Twig\ExpressionParser\ExpressionParserType;
 use Twig\ExpressionParser\InfixExpressionParserInterface;
 use Twig\ExpressionParser\Prefix\LiteralExpressionParser;
 use Twig\ExpressionParser\PrefixExpressionParserInterface;
@@ -59,9 +58,12 @@
     private $ignoreUnknownTwigCallables = false;
     private ExpressionParsers $parsers;
 
+    private Environment $env;
+
     public function __construct(
-        private Environment $env,
+        Environment $env
     ) {
+        $this->env = $env;
         $this->parsers = $env->getExpressionParsers();
     }
 
@@ -229,7 +231,7 @@
                     $subparser->setParser($this);
                     $node = $subparser->parse($token);
                     if (!$node) {
-                        trigger_deprecation('twig/twig', '3.12', 'Returning "null" from "%s" is deprecated and forbidden by "TokenParserInterface".', $subparser::class);
+                        trigger_deprecation('twig/twig', '3.12', 'Returning "null" from "%s" is deprecated and forbidden by "TokenParserInterface".', get_class($subparser));
                     } else {
                         $node->setNodeTag($subparser->getTag());
                         $rv[] = $node;
@@ -330,10 +332,10 @@
         $this->embeddedTemplates[] = $template;
     }
 
-    public function addImportedSymbol(string $type, string $alias, ?string $name = null, AbstractExpression|AssignTemplateVariable|null $internalRef = null): void
+    public function addImportedSymbol(string $type, string $alias, ?string $name = null, /*AbstractExpression|AssignTemplateVariable|null*/ $internalRef = null): void
     {
         if ($internalRef && !$internalRef instanceof AssignTemplateVariable) {
-            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance as an internal reference is deprecated ("%s" given).', __METHOD__, AssignTemplateVariable::class, $internalRef::class);
+            trigger_deprecation('twig/twig', '3.15', 'Not passing a "%s" instance as an internal reference is deprecated ("%s" given).', __METHOD__, AssignTemplateVariable::class, get_class($internalRef));
 
             $internalRef = new AssignTemplateVariable(new TemplateVariable($internalRef->getAttribute('name'), $internalRef->getTemplateLine()), $internalRef->getAttribute('global'));
         }
@@ -423,7 +425,7 @@
         }
 
         if (null !== $parent && !$parent instanceof AbstractExpression) {
-            trigger_deprecation('twig/twig', '3.24', 'Passing a "%s" instance to "%s()" is deprecated, pass an "AbstractExpression" instance instead.', $parent::class, __METHOD__);
+            trigger_deprecation('twig/twig', '3.24', 'Passing a "%s" instance to "%s()" is deprecated, pass an "AbstractExpression" instance instead.', get_class($parent), __METHOD__);
         }
 
         if (null !== $this->parent) {
@@ -588,7 +590,7 @@
         // here, $nested means "being at the root level of a child template"
         // we need to discard the wrapping "Node" for the "body" node
         // Node::class !== \get_class($node) should be removed in Twig 4.0
-        $nested = $nested || (Node::class !== $node::class && !$node instanceof Nodes);
+        $nested = $nested || (Node::class !== get_class($node) && !$node instanceof Nodes);
         foreach ($node as $k => $n) {
             if (null !== $n && null === $this->filterBodyNodes($n, $nested)) {
                 $node->removeNode($k);
@@ -621,7 +623,15 @@
                 }
                 if (isset($this->expressionRefs[$node]) && $ep === $this->expressionRefs[$node]) {
                     $change = $expressionParser->getPrecedenceChange();
-                    trigger_deprecation($change->getPackage(), $change->getVersion(), \sprintf('As the "%s" %s operator will change its precedence in the next major version, add explicit parentheses to avoid behavior change in "%s" at line %d.', $expressionParser->getName(), ExpressionParserType::getType($expressionParser)->value, $this->getStream()->getSourceContext()->getName(), $node->getTemplateLine()));
+
+                    if ($expressionParser instanceof PrefixExpressionParserInterface) {
+                        $type = 'prefix';
+                    }
+                    if ($expressionParser instanceof InfixExpressionParserInterface) {
+                        $type = 'infix';
+                    }
+
+                    trigger_deprecation($change->getPackage(), $change->getVersion(), \sprintf('As the "%s" %s operator will change its precedence in the next major version, add explicit parentheses to avoid behavior change in "%s" at line %d.', $expressionParser->getName(), $type, $this->getStream()->getSourceContext()->getName(), $node->getTemplateLine()));
                 }
             }
         }
@@ -631,7 +641,15 @@
                 /** @var AbstractExpression $node */
                 if (isset($this->expressionRefs[$node]) && $ep === $this->expressionRefs[$node] && !$node->hasExplicitParentheses()) {
                     $change = $ep->getPrecedenceChange();
-                    trigger_deprecation($change->getPackage(), $change->getVersion(), \sprintf('As the "%s" %s operator will change its precedence in the next major version, add explicit parentheses to avoid behavior change in "%s" at line %d.', $ep->getName(), ExpressionParserType::getType($ep)->value, $this->getStream()->getSourceContext()->getName(), $node->getTemplateLine()));
+
+                    if ($ep instanceof PrefixExpressionParserInterface) {
+                        $type = 'prefix';
+                    }
+                    if ($ep instanceof InfixExpressionParserInterface) {
+                        $type = 'infix';
+                    }
+
+                    trigger_deprecation($change->getPackage(), $change->getVersion(), \sprintf('As the "%s" %s operator will change its precedence in the next major version, add explicit parentheses to avoid behavior change in "%s" at line %d.', $ep->getName(), $type, $this->getStream()->getSourceContext()->getName(), $node->getTemplateLine()));
                 }
             }
         }
diff -u a/src/Profiler/Profile.php b/src/Profiler/Profile.php
--- a/src/Profiler/Profile.php
+++ b/src/Profiler/Profile.php
@@ -24,11 +24,17 @@
     private $ends = [];
     private $profiles = [];
 
+    private string $template;
+    private string $type;
+    private string $name;
+
     public function __construct(
-        private string $template = 'main',
-        private string $type = self::ROOT,
-        private string $name = 'main',
+        string $template = 'main',
+        string $type = self::ROOT,
+        string $name = 'main'
     ) {
+        $this->template = $template;
+        $this->type = $type;
         $this->name = str_starts_with($name, '__internal_') ? 'INTERNAL' : $name;
         $this->enter();
     }
diff -u a/src/Runtime/EscaperRuntime.php b/src/Runtime/EscaperRuntime.php
--- a/src/Runtime/EscaperRuntime.php
+++ b/src/Runtime/EscaperRuntime.php
@@ -26,9 +26,12 @@
     /** @internal */
     public $safeLookup = [];
 
+    private $charset;
+
     public function __construct(
-        private $charset = 'UTF-8',
+        $charset = 'UTF-8'
     ) {
+        $this->charset = $charset;
     }
 
     /**
@@ -106,7 +109,7 @@
         if (!\is_string($string)) {
             if ($string instanceof \Stringable) {
                 if ($autoescape) {
-                    $c = $string::class;
+                    $c = get_class($string);
                     if (!isset($this->safeClasses[$c])) {
                         $this->safeClasses[$c] = [];
                         foreach (class_parents($string) + class_implements($string) as $class) {
@@ -199,16 +202,32 @@
                     * Escape sequences supported only by JavaScript, not JSON, are omitted.
                     * \" is also supported but omitted, because the resulting string is not HTML safe.
                     */
-                    $short = match ($char) {
-                        '\\' => '\\\\',
-                        '/' => '\\/',
-                        "\x08" => '\b',
-                        "\x0C" => '\f',
-                        "\x0A" => '\n',
-                        "\x0D" => '\r',
-                        "\x09" => '\t',
-                        default => false,
-                    };
+                    switch ($char) {
+                        case '\\':
+                            $short = '\\\\';
+                            break;
+                        case '/':
+                            $short = '\\/';
+                            break;
+                        case "\x08":
+                            $short = '\b';
+                            break;
+                        case "\x0C":
+                            $short = '\f';
+                            break;
+                        case "\x0A":
+                            $short = '\n';
+                            break;
+                        case "\x0D":
+                            $short = '\r';
+                            break;
+                        case "\x09":
+                            $short = '\t';
+                            break;
+                        default:
+                            $short = false;
+                            break;
+                    }
 
                     if ($short) {
                         return $short;
@@ -265,10 +284,13 @@
                     throw new RuntimeError('The string to escape is not a valid UTF-8 string.');
                 }
 
-                $regex = match ($strategy) {
-                    'html_attr' => '#[^a-zA-Z0-9,\.\-_]#Su',
-                    'html_attr_relaxed' => '#[^a-zA-Z0-9,\.\-_:@\[\]]#Su',
-                };
+                switch ($strategy) {
+                    case 'html_attr':
+                        $regex = '#[^a-zA-Z0-9,\.\-_]#Su';
+                    case 'html_attr_relaxed':
+                        $regex = '#[^a-zA-Z0-9,\.\-_:@\[\]]#Su';
+                        break;
+                }
 
                 $string = preg_replace_callback($regex, static function ($matches) {
                     /**
@@ -299,13 +321,18 @@
                         * entities that XML supports. Using HTML entities would result in this error:
                         *     XML Parsing Error: undefined entity
                         */
-                        return match ($ord) {
-                            34 => '&quot;', /* quotation mark */
-                            38 => '&amp;',  /* ampersand */
-                            60 => '&lt;',   /* less-than sign */
-                            62 => '&gt;',   /* greater-than sign */
-                            default => \sprintf('&#x%02X;', $ord),
-                        };
+                        switch ($ord) {
+                            case 34:
+                                return '&quot;'; /* quotation mark */
+                            case 38:
+                                return '&amp;'; /* ampersand */
+                            case 60:
+                                return '&lt;'; /* less-than sign */
+                            case 62:
+                                return '&gt;'; /* greater-than sign */
+                            default:
+                                return \sprintf('&#x%02X;', $ord);
+                        }
                     }
 
                     /*
diff -u a/src/RuntimeLoader/ContainerRuntimeLoader.php b/src/RuntimeLoader/ContainerRuntimeLoader.php
--- a/src/RuntimeLoader/ContainerRuntimeLoader.php
+++ b/src/RuntimeLoader/ContainerRuntimeLoader.php
@@ -23,9 +23,12 @@
  */
 class ContainerRuntimeLoader implements RuntimeLoaderInterface
 {
+    private ContainerInterface $container;
+
     public function __construct(
-        private ContainerInterface $container,
+        ContainerInterface $container
     ) {
+        $this->container = $container;
     }
 
     public function load(string $class)
diff -u a/src/RuntimeLoader/FactoryRuntimeLoader.php b/src/RuntimeLoader/FactoryRuntimeLoader.php
--- a/src/RuntimeLoader/FactoryRuntimeLoader.php
+++ b/src/RuntimeLoader/FactoryRuntimeLoader.php
@@ -18,12 +18,15 @@
  */
 class FactoryRuntimeLoader implements RuntimeLoaderInterface
 {
+    private array $map;
+
     /**
      * @param array $map An array where keys are class names and values factory callables
      */
     public function __construct(
-        private array $map = [],
+        array $map = []
     ) {
+        $this->map = $map;
     }
 
     public function load(string $class)
diff -u a/src/Sandbox/SecurityPolicy.php b/src/Sandbox/SecurityPolicy.php
--- a/src/Sandbox/SecurityPolicy.php
+++ b/src/Sandbox/SecurityPolicy.php
@@ -130,7 +130,7 @@
         }
 
         if (!$allowed) {
-            $class = $obj::class;
+            $class = get_class($obj);
             throw new SecurityNotAllowedMethodError(\sprintf('Calling "%s" method on a "%s" object is not allowed.', $method, $class), $class, $method);
         }
     }
@@ -146,7 +146,7 @@
         }
 
         if (!$allowed) {
-            $class = $obj::class;
+            $class = get_class($obj);
             throw new SecurityNotAllowedPropertyError(\sprintf('Calling "%s" property on a "%s" object is not allowed.', $property, $class), $class, $property);
         }
     }
diff -u a/src/Source.php b/src/Source.php
--- a/src/Source.php
+++ b/src/Source.php
@@ -18,16 +18,23 @@
  */
 final class Source
 {
+    private string $code;
+    private string $name;
+    private string $path;
+
     /**
      * @param string $code The template source code
      * @param string $name The template logical name
      * @param string $path The filesystem path of the template if any
      */
     public function __construct(
-        private string $code,
-        private string $name,
-        private string $path = '',
+        string $code,
+        string $name,
+        string $path = ''
     ) {
+        $this->code = $code;
+        $this->name = $name;
+        $this->path = $path;
     }
 
     public function getCode(): string
diff -u a/src/Template.php b/src/Template.php
--- a/src/Template.php
+++ b/src/Template.php
@@ -42,9 +42,12 @@
 
     private $useYield;
 
+    protected Environment $env;
+
     public function __construct(
-        protected Environment $env,
+        Environment $env
     ) {
+        $this->env = $env;
         $this->useYield = $env->useYield();
         $this->extensions = $env->getExtensions();
     }
@@ -74,7 +77,7 @@
      *
      * @return self|TemplateWrapper|false The parent template or false if there is no parent
      */
-    public function getParent(array $context): self|TemplateWrapper|false
+    public function getParent(array $context)/*: self|TemplateWrapper|false*/
     {
         if (null !== $this->parent) {
             return $this->parent;
@@ -103,7 +106,7 @@
         return $this->parents[$parent];
     }
 
-    protected function doGetParent(array $context): bool|string|self|TemplateWrapper
+    protected function doGetParent(array $context)/*: bool|string|self|TemplateWrapper*/
     {
         return false;
     }
@@ -278,7 +281,7 @@
     /**
      * @param string|TemplateWrapper|array<string|TemplateWrapper> $template
      */
-    protected function load(string|TemplateWrapper|array $template, int $line, ?int $index = null): self
+    protected function load(/*string|TemplateWrapper|array*/ $template, int $line, ?int $index = null): self
     {
         try {
             if (\is_array($template)) {
@@ -323,7 +326,7 @@
      *
      * @deprecated since Twig 3.21 and will be removed in 4.0. Use Template::load() instead.
      */
-    protected function loadTemplate($template, $templateName = null, ?int $line = null, ?int $index = null): self|TemplateWrapper
+    protected function loadTemplate($template, $templateName = null, ?int $line = null, ?int $index = null)/*: self|TemplateWrapper*/
     {
         trigger_deprecation('twig/twig', '3.21', 'The "%s" method is deprecated.', __METHOD__);
 
diff -u a/src/TemplateWrapper.php b/src/TemplateWrapper.php
--- a/src/TemplateWrapper.php
+++ b/src/TemplateWrapper.php
@@ -18,6 +18,9 @@
  */
 final class TemplateWrapper
 {
+    private Environment $env;
+    private Template $template;
+
     /**
      * This method is for internal use only and should never be called
      * directly (use Twig\Environment::load() instead).
@@ -25,9 +28,11 @@
      * @internal
      */
     public function __construct(
-        private Environment $env,
-        private Template $template,
+        Environment $env,
+        Template $template
     ) {
+        $this->env = $env;
+        $this->template = $template;
     }
 
     /**
diff -u a/src/Test/IntegrationTestCase.php b/src/Test/IntegrationTestCase.php
--- a/src/Test/IntegrationTestCase.php
+++ b/src/Test/IntegrationTestCase.php
@@ -287,14 +287,14 @@
             } catch (\Exception $e) {
                 if (false !== $exception) {
                     $message = $e->getMessage();
-                    $this->assertSame(trim($exception), trim(\sprintf('%s: %s', $e::class, $message)));
+                    $this->assertSame(trim($exception), trim(\sprintf('%s: %s', get_class($e), $message)));
                     $last = substr($message, \strlen($message) - 1);
                     $this->assertTrue('.' === $last || '?' === $last, 'Exception message must end with a dot or a question mark.');
 
                     return;
                 }
 
-                throw new Error(\sprintf('%s: %s', $e::class, $e->getMessage()), -1, null, $e);
+                throw new Error(\sprintf('%s: %s', get_class($e), $e->getMessage()), -1, null, $e);
             } finally {
                 restore_error_handler();
             }
@@ -305,14 +305,14 @@
                 $output = trim($template->render(eval($match[1].';')), "\n ");
             } catch (\Exception $e) {
                 if (false !== $exception) {
-                    $this->assertStringMatchesFormat(trim($exception), trim(\sprintf('%s: %s', $e::class, $e->getMessage())));
+                    $this->assertStringMatchesFormat(trim($exception), trim(\sprintf('%s: %s', get_class($e), $e->getMessage())));
 
                     return;
                 }
 
-                $e = new Error(\sprintf('%s: %s', $e::class, $e->getMessage()), -1, null, $e);
+                $e = new Error(\sprintf('%s: %s', get_class($e), $e->getMessage()), -1, null, $e);
 
-                $output = trim(\sprintf('%s: %s', $e::class, $e->getMessage()));
+                $output = trim(\sprintf('%s: %s', get_class($e), $e->getMessage()));
             }
 
             if (false !== $exception) {
diff -u a/src/TokenParser/GuardTokenParser.php b/src/TokenParser/GuardTokenParser.php
--- a/src/TokenParser/GuardTokenParser.php
+++ b/src/TokenParser/GuardTokenParser.php
@@ -41,7 +41,7 @@
 
         try {
             $exists = null !== $this->parser->getEnvironment()->$method($name);
-        } catch (SyntaxError) {
+        } catch (SyntaxError $e) {
             $exists = false;
         }
 
diff -u a/src/Token.php b/src/Token.php
--- a/src/Token.php
+++ b/src/Token.php
@@ -39,11 +39,19 @@
      */
     public const SPREAD_TYPE = 13;
 
+    private int $type;
+    private $value;
+    private int $lineno;
+
     public function __construct(
-        private int $type,
-        private $value,
-        private int $lineno,
+        int $type,
+        $value,
+        int $lineno
     ) {
+        $this->type = $type;
+        $this->value = $value;
+        $this->lineno = $lineno;
+
         if (self::ARROW_TYPE === $type) {
             trigger_deprecation('twig/twig', '3.21', 'The "%s" token type is deprecated, "arrow" is now an operator.', self::ARROW_TYPE);
         }
diff -u a/src/TokenStream.php b/src/TokenStream.php
--- a/src/TokenStream.php
+++ b/src/TokenStream.php
@@ -22,11 +22,16 @@
 final class TokenStream
 {
     private $current = 0;
+    private array $tokens;
+    private ?Source $source;
 
     public function __construct(
-        private array $tokens,
-        private ?Source $source = null,
+        array $tokens,
+        ?Source $source = null
     ) {
+        $this->tokens = $tokens;
+        $this->source = $source;
+
         if (null === $this->source) {
             trigger_deprecation('twig/twig', '3.16', \sprintf('Not passing a "%s" object to "%s" constructor is deprecated.', Source::class, __CLASS__));
 
diff -u a/src/Util/CallableArgumentsExtractor.php b/src/Util/CallableArgumentsExtractor.php
--- a/src/Util/CallableArgumentsExtractor.php
+++ b/src/Util/CallableArgumentsExtractor.php
@@ -28,10 +28,16 @@
 {
     private ReflectionCallable $rc;
 
+    private Node $node;
+    private TwigCallableInterface $twigCallable;
+
     public function __construct(
-        private Node $node,
-        private TwigCallableInterface $twigCallable,
+        Node $node,
+        TwigCallableInterface $twigCallable
     ) {
+        $this->node = $node;
+        $this->twigCallable = $twigCallable;
+
         $this->rc = new ReflectionCallable($twigCallable);
     }
 
diff -u a/src/Util/DeprecationCollector.php b/src/Util/DeprecationCollector.php
--- a/src/Util/DeprecationCollector.php
+++ b/src/Util/DeprecationCollector.php
@@ -20,9 +20,12 @@
  */
 final class DeprecationCollector
 {
+    private Environment $twig;
+
     public function __construct(
-        private Environment $twig,
+        Environment $twig
     ) {
+        $this->twig = $twig;
     }
 
     /**
diff -u a/src/Util/ReflectionCallable.php b/src/Util/ReflectionCallable.php
--- a/src/Util/ReflectionCallable.php
+++ b/src/Util/ReflectionCallable.php
@@ -25,7 +25,7 @@
     private $name;
 
     public function __construct(
-        TwigCallableInterface $twigCallable,
+        TwigCallableInterface $twigCallable
     ) {
         $callable = $twigCallable->getCallable();
         if (\is_string($callable) && false !== $pos = strpos($callable, '::')) {
diff -u a/src/ExpressionParser/ExpressionParserType.php b/src/ExpressionParser/ExpressionParserType.php
deleted file mode 100644
--- a/src/ExpressionParser/ExpressionParserType.php
+++ /dev/null
@@ -1,33 +0,0 @@
-<?php
-
-/*
- * This file is part of Twig.
- *
- * (c) Fabien Potencier
- *
- * For the full copyright and license information, please view the LICENSE
- * file that was distributed with this source code.
- */
-
-namespace Twig\ExpressionParser;
-
-/**
- * @internal
- */
-enum ExpressionParserType: string
-{
-    case Prefix = 'prefix';
-    case Infix = 'infix';
-
-    public static function getType(object $object): ExpressionParserType
-    {
-        if ($object instanceof PrefixExpressionParserInterface) {
-            return self::Prefix;
-        }
-        if ($object instanceof InfixExpressionParserInterface) {
-            return self::Infix;
-        }
-
-        throw new \InvalidArgumentException(\sprintf('Unsupported expression parser type: %s', $object::class));
-    }
-}
diff -u a/src/ExpressionParser/InfixAssociativity.php b/src/ExpressionParser/InfixAssociativity.php
deleted file mode 100644
--- a/src/ExpressionParser/InfixAssociativity.php
+++ /dev/null
@@ -1,18 +0,0 @@
-<?php
-
-/*
- * This file is part of Twig.
- *
- * (c) Fabien Potencier
- *
- * For the full copyright and license information, please view the LICENSE
- * file that was distributed with this source code.
- */
-
-namespace Twig\ExpressionParser;
-
-enum InfixAssociativity
-{
-    case Left;
-    case Right;
-}
diff -u a/src/Node/MacroNode.php b/src/Node/MacroNode.php
--- a/src/Node/MacroNode.php
+++ b/src/Node/MacroNode.php
@@ -78,7 +78,7 @@
 
         $compiler
             ->raw('...$varargs')
-            ->raw("): string|Markup\n")
+            ->raw(")/*: string|Markup*/\n")
             ->write("{\n")
             ->indent()
             ->write("\$macros = \$this->macros;\n")
diff -u a/src/Node/ModuleNode.php b/src/Node/ModuleNode.php
--- a/src/Node/ModuleNode.php
+++ b/src/Node/ModuleNode.php
@@ -139,7 +139,7 @@
         $parent = $this->getNode('parent');
 
         $compiler
-            ->write("protected function doGetParent(array \$context): bool|string|Template|TemplateWrapper\n", "{\n")
+            ->write("protected function doGetParent(array \$context)/*: bool|string|Template|TemplateWrapper*/\n", "{\n")
             ->indent()
             ->addDebugInfo($parent)
             ->write('return ')
