diff --git a/src/Drivers/SplitbrainPhpArchive.php b/src/Drivers/SplitbrainPhpArchive.php
index 7ef3677..1373ce8 100644
--- a/src/Drivers/SplitbrainPhpArchive.php
+++ b/src/Drivers/SplitbrainPhpArchive.php
@@ -74,8 +74,8 @@ class SplitbrainPhpArchive extends BasicPureDriver
         switch ($format) {
             case Formats::ZIP:
             case Formats::TAR:
-            case Formats::TAR_GZIP;
-            case Formats::TAR_BZIP;
+            case Formats::TAR_GZIP:
+            case Formats::TAR_BZIP:
                 return [
                     Abilities::OPEN,
 //                    BasicDriver::EXTRACT_CONTENT,
