diff --git a/src/Drivers/OneFile/OneFileDriver.php b/src/Drivers/OneFile/OneFileDriver.php
index fee038b..96a1e00 100644
--- a/src/Drivers/OneFile/OneFileDriver.php
+++ b/src/Drivers/OneFile/OneFileDriver.php
@@ -110,7 +110,7 @@ public function getFileData($fileName)
      * @return int
      * @throws ArchiveExtractionException
      */
-    public function extractFiles($outputFolder, array $files = null)
+    public function extractFiles($outputFolder, ?array $files = null)
     {
         return $this->extractArchive($outputFolder);
     }
