diff --git a/src/Reservation.php b/src/Reservation.php
index 2f9c27e08b..3b25289270 100644
--- a/src/Reservation.php
+++ b/src/Reservation.php
@@ -187,10 +187,7 @@ class Reservation extends CommonDBChild
         if (empty($input['users_id'])) {
             $input['users_id'] = Session::getLoginUserID();
         }
-        if (
-            !Session::haveRight("reservation", UPDATE)
-            && Session::getLoginUserID() !== $input['users_id']
-        ) {
+        if (!Session::haveRight("reservation", ReservationItem::RESERVEANITEM)) {
             return;
         }
 
