diff -up ./src/Handler/CurlFactory.php.patch ./src/Handler/CurlFactory.php
--- ./src/Handler/CurlFactory.php.patch	2017-06-07 08:45:31.902697878 +0200
+++ ./src/Handler/CurlFactory.php	2017-06-07 08:45:38.720764102 +0200
@@ -16,7 +16,7 @@ use Psr\Http\Message\RequestInterface;
 class CurlFactory implements CurlFactoryInterface
 {
     /** @var array */
-    private $handles;
+    private $handles = [];
 
     /** @var int Total number of idle handles to keep in cache */
     private $maxHandles;
